### PR TITLE
Make sure config providers are unique

### LINKED ISSUE #36: Validation for `class-string` config providers to be unique
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes
| RFC         | yes
| BC Break    | no

#### Summary

I recently detected some duplicate entries within the config provider array which is passed to the `ConfigAggregator#__construct`. Since it requires the config to be merged multiple times while it makes no sense, I'd say we could add some validation to the `ConfigAggregator#__construct` to at least report duplicated `class-string` and/or `function-string` values.

WDYT?

### PR DESCRIPTION
**Pull request by alexmerlin**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes
| RFC           | no
| QA            | no

### Description

When initializing the `ConfigAggregator`, this will iterate through the provided items and look for duplicates.
If there are duplicates, a `InvalidConfigProviderException` exception is thrown with the message:

> `Some\Provider` is registered more than once. Config providers should be unique. In case a specific order is required, please double check before deleting the duplicate(s).

### CODE REVIEW

File: src/ConfigAggregator.php
Comment (Ocramius) [inline src/ConfigAggregator.php]: I don't think we should serialize the provider: risky operation, and big assumptions around it being serializable

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/ConfigAggregator.php
Comment (Xerkus) [inline src/ConfigAggregator.php]: Object uniqueness can not be assumed from it's class name. Unless it is exactly same instance it must be treated as unique.

File: src/ConfigAggregator.php
Comment (alexmerlin) [reply src/ConfigAggregator.php]: I'm not happy with it either, but I needed a solution to convert to string whatever is stored in `$provider` (knowing that it is not an object nor a string).
Any recommendation?

**Comment by Xerkus** [comment]
I am not sure how useful this is.

Only class-strings could be used to determine unique config providers. Closures or objects must always be treated as unique. 

I do not think duplicated config providers should prevent application from starting.  It is merely a performance concern of having to do some more recursive merging.

File: src/ConfigAggregator.php
Comment (Xerkus) [reply src/ConfigAggregator.php]: This solution introduces more overhead than it tries to solve tbh. This will always have a performance hit while merging some config provider twice is only when config provider appears twice which is virtually never.

File: src/ConfigAggregator.php
Comment (alexmerlin) [reply src/ConfigAggregator.php]: [This comment](https://github.com/laminas/laminas-config-aggregator/issues/36#issuecomment-1710510150) only raised the question of two classes being of the same type (not same instance).

If I understand correctly, you say that we should accept this:
```php
$config = new ConfigAggregator([
    Foo::class, // introduces service interface XY
    Bar::class, // overrides service interface XY with whatever
    Foo::class, // overrides service interface XY with whatever back to the initial service
]);
```
as long as the two occurrences of `Foo` are not the same instance?

**Comment by alexmerlin** [comment]
@Xerkus focusing only on class-strings would definitely simplify this feature - if you say that making sure that there are no duplicate class-strings and ignore other types is enough, I'm fine with that.

File: src/ConfigAggregator.php
Comment (samsonasik) [reply src/ConfigAggregator.php]: I am on mobile, just quick look, is array_unique() with SORT_REGULAR constant should already cover this? ref https://3v4l.org/gDm7T

Comment (alexmerlin) [reply src/ConfigAggregator.php]: My initial plan was to use `array_unique`, but because different types of providers (arrays, class-strings, closures, generators) needed different treatment in the code, I opted for the more complicated solution.
Now, if we agree that this feature should only focus on class-strings, than I can simplify things.

**Comment by Xerkus** [comment]
Only class-string is guaranteed to have same behavior. Objects are not.

I do not need to go far for an example: mezzio skeleton has two instances of `PhpFileProvider`. `ArrayProviders` are also likely to appear several times. Closures are all considered to be instances of the same class. Config provider can be configurable eg as an instance with feature flags. Config provider could be a wrapper with some logic. And so on.

Most common usage where duplicate entry can appear is getting config provider injected via our composer plugin. In that case dealing only with class-strings is perfectly fine.
Composer plugin does not normally inject duplicate entries. They can be introduced via faulty git merge, however.

File: src/ConfigAggregator.php
Comment (Xerkus) [reply src/ConfigAggregator.php]: I thought of array_unique too but duplicate entries would need to be found and listed for this to have a useful message.

**Comment by Xerkus** [comment]
I would like to propose an alternative approach:
implement unique providers validator as a pre-processor that can be passed to the config provider as a constructor argument.

Edit: Actually, nevermind. This would only work if pre-processor is the first to be applied and it would be relying on implementation detail of config aggregator not instantiating  providers before passing to pre-processors.

File: src/ConfigAggregator.php
Comment (boesing) [reply src/ConfigAggregator.php]: Use `get_debug_type` instead, should be good to go.
An additional text like `A <whatever config provider type> is already registered` or sth. like that.
But I highly doubt that two callables can be checked via `in_array` anyways and thus, you might only end-up with `class-string` or instances of a compatible config provider. Though `get_debug_type` might be good to go here.

File: src/ConfigAggregator.php
Comment (Xerkus) [reply src/ConfigAggregator.php]: > If I understand correctly, you say that we should accept this:

No, that is exactly the case that can be safely detected and reported.


This, however, would be requiring a comparison of internal state which is a no go, really. This should be accepted:

```php
$config = new ConfigAggregator([
    new Foo(), // introduces service interface XY
    Bar::class, // overrides service interface XY with whatever
    new Foo(), // overrides service interface XY with whatever back to the initial service
]);
```

**Comment by boesing** [comment]
I'd focus only on `class-string` as outlined in my initial issue, *maybe* add strict comparison to prevent having multiple `exact` duplications as well (this would include both `callable` and `object` maybe).

Add some tests to verify the behavior. Might be even worth to have `!is_string() && !is_object()` check to skip duplication verification.

File: src/ConfigAggregator.php
Comment (Xerkus) [reply src/ConfigAggregator.php]: objects or even closures can be reported if `===` is used but I can't imagine same instance appearing twice outside of tests.

**Comment by alexmerlin** [comment]
To summarize, we should go with:

- checking only `class-strings` (anything where `get_debug_type($provider)` returns `string`)
- iterate over `$providers` (instead of using `array_unique`) and check if a `$provider` is already registered in `$uniqueProviders`
- on duplicate, throw `InvalidConfigProviderException` with message `A is already registered`

All agree on the above?

File: src/ConfigAggregator.php
Comment (boesing) [reply src/ConfigAggregator.php]: I also do not expect instances being duplicated, but I've seen it a bunch of times where `class-string` was.

**Comment by boesing** [comment]
> To summarize, we should go with:
> 
> * checking only `class-strings` (anything where `get_debug_type($provider)` returns `string`)
> * iterate over `$providers` (instead of using `array_unique`) and check if a `$provider` is already registered in `$uniqueProviders`
> * on duplicate, throw `InvalidConfigProviderException` with message `A is already registered`
> 
> All agree on the above?

I'd rephrase this a bit, but ya:

> * iterate over `$providers` (instead of using `array_unique`) and check if a `$provider` is already registered in `$uniqueProviders` using strict comparison
> * on duplicate, throw `InvalidConfigProviderException` with message `<debug type> is registered more than once. Config providers should be unique. In case a specific order is required, please double check before deleting the duplicate(s).` (something like this)

**Comment by Xerkus** [comment]
I would say, sort array then iterate and check if previous entry is the same. Assemble a list of duplicate entries for the  error message.

File: src/ConfigAggregator.php
Comment (samsonasik) [inline src/ConfigAggregator.php]: the method is actually validating, not make unique, so method name maybe need change, eg: `verifyDuplicateProviders` or `validateDuplicateProviders`

Comment (alexmerlin) [reply src/ConfigAggregator.php]: Indeed, with the most recent changes, the name `removeDuplicateProviders` is not the best fit anymore.
I guess `validateDuplicateProviders` should do it.

Pros/cons anybody?

File: test/ConfigAggregatorTest.php
Comment (Xerkus) [inline test/ConfigAggregatorTest.php]: Wait. This works? More instances could be created for anonymous class using generated class name?

Comment (alexmerlin) [reply test/ConfigAggregatorTest.php]: yup

File: src/ConfigAggregator.php
Comment (boesing) [inline src/ConfigAggregator.php]: `get_debug_type` will be always string.

```suggestion
            if (!is_string($provider) && !is_object($provider)) {
                continue;
            }
            
            if (in_array($provider, $uniqueProviders, true)) {
```

Comment (alexmerlin) [reply src/ConfigAggregator.php]: Always string?
For instances of an anonymous class it returns `class@anonymous`
For instances of a real class it returns the class name
For anonymous functions and generators it returns `Closure`
For arrays it returns `array`
For class-strings it returns `string`

Comment (Xerkus) [reply src/ConfigAggregator.php]: `get_debug_type` is used for printing debug info. If you want to check for string use `is_string()`

File: src/InvalidConfigProviderException.php
Comment (Xerkus) [inline src/InvalidConfigProviderException.php]: declare parameter and return types

Comment (alexmerlin) [reply src/InvalidConfigProviderException.php]: When I created `fromDuplicateProvider`, I followed the structure of the other two methods (`fromNamedProvider` and `fromUnsupportedType`).
Ok, I'll add parameter and return types to this method.

File: src/ConfigAggregator.php
Comment (Xerkus) [reply src/ConfigAggregator.php]: @alexmerlin `get_debug_type` suggested in another comment was for reporting the problematic provider not for conditionals.

For example, when `$provider` can be something other than string we would report it this way:

```php
        return new self(sprintf(
            '%s is registered more than once. Config providers should be unique. In case a specific order is'
            . ' required, please double check before deleting the duplicate(s).',
            is_string($provider) ? $provider : get_debug_type($provider)
        ));
```

File: src/InvalidConfigProviderException.php
Comment (Xerkus) [reply src/InvalidConfigProviderException.php]: A leftover from before the types were added, most likely.

File: src/ConfigAggregator.php
Comment (Xerkus) [inline src/ConfigAggregator.php]: I think it would be better as
```suggestion
    private function validateNoDuplicateProviders(iterable $providers): iterable
```

Comment (alexmerlin) [reply src/ConfigAggregator.php]: Modified.

**Comment by Xerkus** [comment]
Now tests are needed to check:
- instances of the same class are allowed
- two closures are allowed
- same instance twice is not allowed
- duplicate providers with valid providers in-between are still not allowed
- duplicate providers are checked before pre-processors
- ? duplicate providers injected by pre-processors are not checked ?

**Comment by Xerkus** [comment]
And the last thing to add is documentation update for the feature

File: src/ConfigAggregator.php
Comment (boesing) [inline src/ConfigAggregator.php:118]: Not 100% sure but afair `get_debug_type` keeps string if string input is given. Maybe double check that, but I remember something. 🤔

Comment (alexmerlin) [reply src/ConfigAggregator.php:118]: If I use `get_debug_type($provider)` on strings as well, then the error thrown would look something like:

> string is already registered...

which does not provide much info on what went wrong.

With my current implementation it will throw:

> Some\Provider is already registered...

which is much more relevant.

Comment (boesing) [reply src/ConfigAggregator.php:118]: Ah okay, I was expecting that it wont change the string as it is already a string. good to know, thanks for clarification. havent had time to play around with that function and since I mostly write via my phone, I did not tested it 👍🏻

**Comment by alexmerlin** [comment]
Does anyone have anything to add to this?

File: test/ConfigAggregatorTest.php
Comment (gsteel) [inline test/ConfigAggregatorTest.php]:
```suggestion
```

This is unreachable

File: test/ConfigAggregatorTest.php
Comment (gsteel) [inline test/ConfigAggregatorTest.php]:
```suggestion
```

Unreachable

File: src/ConfigAggregator.php
Comment (gsteel) [inline src/ConfigAggregator.php:113]: Because you are returning `$uniqueProviders`, this method is also effectively filtering entirely invalid entries too, so `[123, Some::class]` would become `[Some::class]` - Obviously `123` is invalid, but are you supressing otherwise noisy errors for invalid input?

Not suggesting this _must_ be changed, but it's worth considering the implications. Might be easier to just return the un-modified input

**Review by gsteel — CHANGES_REQUESTED** [review]

File: src/ConfigAggregator.php
Comment (Xerkus) [reply src/ConfigAggregator.php:113]: This explicitly filters out valid provides I mentioned exist.

**Comment by Xerkus** [comment]
This should not modify providers in any way and only assert no duplicates exist. Ie fail - no fail.

**Comment by Xerkus** [comment]
My request for test cases is not addressed. My bad, I should have done it as part of the submitted review requesting change.

**Comment by alexmerlin** [comment]
@gsteel @Xerkus have another look when you can

File: test/ConfigAggregatorTest.php
Comment (alexmerlin) [reply test/ConfigAggregatorTest.php]: removed

File: test/ConfigAggregatorTest.php
Comment (alexmerlin) [reply test/ConfigAggregatorTest.php]: removed

File: src/ConfigAggregator.php
Comment (alexmerlin) [reply src/ConfigAggregator.php:113]: Modified method, so that it does not modify the providers.

File: src/ConfigAggregator.php
Comment (gsteel) [inline src/ConfigAggregator.php:106]: I'd prefer this method to be named `assertSomething` rather than `validateSomething` because it is an assertion that will throw an exception as opposed to a validation that would typically return success or failure - Just a nitpick and not a blocker for me.

**Review by gsteel — APPROVED** [review]
LGTM, thanks @alexmerlin - I'll wait for review from @Xerkus before merging/tagging

File: src/ConfigAggregator.php
Comment (alexmerlin) [reply src/ConfigAggregator.php:106]: `validateNoDuplicateProviders` -> `assertNoDuplicateProviders`?

Comment (gsteel) [reply src/ConfigAggregator.php:106]: I'm not _that_ picky, but yeah, something like that, or `assertUniqueConfigProviders`?

As an assertion, you could drop the return value entirely, i.e. `return void`

Comment (alexmerlin) [reply src/ConfigAggregator.php:106]: Btw, the current method naming comes from [this conversation](https://github.com/laminas/laminas-config-aggregator/pull/47/commits/5182755b9bc7e747bcc714560e025ea7f8919af6#r1709310928) with @samsonasik and later [this conversation](https://github.com/laminas/laminas-config-aggregator/pull/47/commits/ccec0e4c82f6f896d052726a6ec2120b31ded642#r1709519256) with @Xerkus

Comment (gsteel) [reply src/ConfigAggregator.php:106]: 😂 - No worries! Like I said - it's not a blocker for me 👍

**Comment by alexmerlin** [comment]
@Xerkus Any input on this?

**Review by samsonasik — APPROVED** [review]

**Pull request merged**
